kw6/reader.py

```diff
@@ -20,61 +20,118 @@ class Reader(BaseModel):
         from pathlib import Path
         import kw6
 
-        path = Path('...')
+        path = Path("...")
 
         for position in kw6.Reader.from_path(path):
             for camera in position.cameras:
                 camera.image.save(
-                    f'{position.header.frame_index}_{camera.header.camera_index}.png'
+                    f"{position.header.frame_index}_{camera.header.camera_index}.png"
                 )
-    '''
+    """
 
+    path: Path
     stream: io.BufferedReader
     cached_positions: Dict[int, int]
-    n_bytes_position: int
     initial_frame_index: int
     inconsistent_position_bytes: bool = False
 
     class Config:
         arbitrary_types_allowed = True
 
     @staticmethod
-    def from_path(path, header_path=None):
-        stream = Path(path).open('rb')
+    @validate_arguments
+    def from_path(path: Path, header_path: Path = None):
+        stream = io.BufferedReader(path.open("rb"))
+
         version = stream.read(settings.N_BYTES_VERSION).decode().strip()
-        if version != 'KW6FileClassVer1.0':
-            raise ValueError(f'Unexpected file version {version}')
-        n_bytes_position, initial_frame_index = initial_position_info(stream)
+        if version != "KW6FileClassVer1.0":
+            raise ValueError(f"Unexpected file version {version}")
+
+        initial_position_header = Position.skip_(stream)
         return Reader(
+            path=path,
             stream=stream,
-            cached_positions=header_positions(header_path) if header_path is not None else {},
-            n_bytes_position=n_bytes_position,
-            initial_frame_index=initial_frame_index,
+            cached_positions=(
+                header_positions(header_path)
+                if header_path is not None
+                else dict()
+            ),
+            initial_frame_index=initial_position_header.frame_index,
         )
 
+    def __iter__(self):
+        """Iterate over all positions and cameras in the file"""
+        stream = io.BufferedReader(self.path.open("rb"))
+        stream.seek(settings.N_BYTES_VERSION)
+        while not stream.peek(1) == b"":
+            byte_position = stream.tell()
+            position = Position.from_stream_(stream)
+            self.cached_positions[position.header.frame_index] = byte_position
+            yield position
+        stream.close()
+
+    def __len__(self):
+        try:
+            return self.assumptuous_length()
+        except Exception:
+            self.stream.seek(settings.N_BYTES_VERSION)
+            if self.empty():
+                return 0
+
+            position_header = None
+            while not self.empty():
```

Comment:
Did you try how long this takes on a normal length kw6 file? You dont expect `len(object)` to be an expensive operation usually

Reply:
0.3 s for 3.5 GB